ufork-wasm/lib/cell.asm

```diff
@@ -20,48 +26,131 @@
 ;;      END
 ;;  ]
 ;;  CREATE cell WITH cell_beh(0)
-cell_beh:               ; (value) <- (tag cust . req)
+beh:                    ; (value) <- (tag cust . req)
     msg 1               ; tag
-    eq 0                ; tag==read
-    if cell_read        ; --
+    eq read_tag         ; tag==read
+    if read             ; --
     msg 1               ; tag
-    eq 1                ; tag==write
-    if cell_write       ; --
+    eq write_tag        ; tag==write
+    if write            ; --
     msg 1               ; tag
-    eq 2                ; tag==CAS
-    if cell_CAS         ; --
+    eq CAS_tag          ; tag==CAS
+    if CAS              ; --
     ref std.abort
 
-cell_read:
+read:                   ; <- (tag cust)
     state 1             ; value
     msg 2               ; value cust
     ref std.send_msg
 
-cell_write:
+write:                  ; <- (tag cust value')
     msg -2              ; (value')
     my beh              ; (value') beh
     beh -1              ; --
     my self             ; SELF
     msg 2               ; SELF cust
     ref std.send_msg
 
-cell_CAS:
+CAS:                    ; <- (tag cust old new)
     msg 3               ; old
     state 1             ; old value
     cmp eq              ; old==value
-    if_not cell_read    ; --
-    msg -3              ; (new)
-    my beh              ; (new) beh
-    beh -1              ; --
-    ref cell_read
+    if_not read         ; --
+    msg 4               ; new
```

Comment:
The instruction sequence
```
    msg -3              ; (new)
    my beh              ; (new) beh
    beh -1              ; --
```
is equivalent to
```
    msg 4               ; new
    my beh              ; new beh
    beh 1               ; --
```
however, the first form re-uses a cell directly from the message instead of creating a new copy on the stack.

Reply:
Oh right, I will revert my change in my next PR. I just found it a bit confusing, but that is because I haven't had much practice with the negative indexing.